VoodooPS2Trackpad/VoodooPS2SynapticsTouchPad.cpp

```diff
@@ -1576,8 +1589,10 @@ void ApplePS2SynapticsTouchPad::sendTouchData() {
         super::messageClient(kIOMessageVoodooInputUpdateDimensionsMessage, voodooInputInstance, &d, sizeof(VoodooInputDimensions));
     }
 
-    // send the event into the multitouch interface
-    super::messageClient(kIOMessageVoodooInputMessage, voodooInputInstance, &inputEvent, sizeof(VoodooInputEvent));
+    if (inputEvent.contact_count) {
```

Comment:
This looks bad. If in the middle of four-finger swipe all fingers disappear, we'll not notify VoodooInput and so the OS will think that the fingers are still on the touchpad, so the gesture will get stuck. That's what I fixed first, when grabbed MT2 simulation from kprinssu's repository.

Reply:
The code from L1569 to 1581 detects finger release and sends the "finger leaving" message, in which `transducer.isTransducerActive` is set to 0, to tell VoodooInput the fingers are leaving.
Therefore, it is okay to stop VoodooPS2 periodically sending empty or old messages here.

Reply:
It even seems to be the opposite of what you say in the description.

Reply:
> It even seems to be the opposite of what you say in the description.

Would you mind elaborating a bit further? What's the test result?

The part above just simulates a one-off state change for leaving fingers in VoodooI2C's way.
I2C-HID touchpad sends finger data with a touch switch state included, which is necessary for finger state tracking.

[alexandred/VoodooI2CHID/VoodooI2CHID/VoodooI2CMultitouchHIDEventDriver.cpp#L270-L274](https://github.com/alexandred/VoodooI2CHID/blob/bce6f5c622e8490f43e3c35e2c848d87189a0253/VoodooI2CHID/VoodooI2CMultitouchHIDEventDriver.cpp#L270-L274)
```cpp
                    case kHIDUsage_Dig_Touch:
                    case kHIDUsage_Dig_TipSwitch:
                        setButtonState(&transducer->tip_switch, 0, value, timestamp);
                        handled    |= element_is_current;
                        break;
```

https://github.com/alexandred/VoodooI2C/blob/158a89338b52cdc47894887944f6eedd99a5ce80/Multitouch%20Support/Native/VoodooI2CNativeEngine.cpp#L37

```cpp
        inputTransducer->isTransducerActive = transducer->tip_switch.value();
```

https://github.com/acidanthera/VoodooInput/blob/845aa78d4f855302157e00fa24149efa828139dc/VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp#L204-L211

```cpp
        if (!transducer->isTransducerActive && !input_report.Button) {
            finger_data.State = 0x7;
            finger_data.Priority = 0x5;
            finger_data.Size = 0x0;
            finger_data.Pressure = 0x0;
            finger_data.Touch_Minor = 0;
            finger_data.Touch_Major = 0;
        }
```

Reply:
@Goshin 
From what I see in this pull request, you do not send any messages if the finger count is 0. So VoodooInput can't get that `0` in `transducer.isTransducerActive`.

Reply:
In which version of file do you reference the line numbers?

Reply:
> @Goshin
> From what I see in this pull request, you do not send any messages if the finger count is 0. So VoodooInput can't get that `0` in `transducer.isTransducerActive`.

Well, at least once in my view,
because
```cpp
        inputEvent.contact_count = last_transducers_count;
```

> In which version of file do you reference the line numbers?

right in the Files Changed tab
https://github.com/acidanthera/VoodooPS2/pull/12/files

Reply:
@Goshin Oh, I missed that line. Well, that may make sense. However, I tested sending the message only once, and:
- Smart Zoom works very rarely (~1 of 40 times);
- sometimes the system doesn't fully react to releasing fingers:
![Снимок экрана 2020-04-09 в 19 19 55](https://user-images.githubusercontent.com/446776/78917360-716f6e80-7a97-11ea-8d25-5e19e5ce520b.png)

Reply:
Yes, it is expected and intended to send only once, in the same way of I2C-HID touchpad.

The smart zoom problem is due to the low report rate of the touchpad. Try this patch on VoodooInput.

https://github.com/Goshin/VoodooInput/commit/02071d3101312220565ae25966ad6745a25c985e

Reply:
Hm, I tested before this PR and fingers are still sometimes left with 0 size, so it seems unrelated. Though without the fix it seems to be completely impossible to trigger Smart Zoom, and with the fix it works at least sometimes. It also fixes a problem in three-finger drag. I reimplemented only this part of this fix in cfa2484. Thanks!

Reply:
For MT2Simulator, the `touch_state` changes are necessary for some gestures, but it needs to reach `touch_state = 4` to be recognized as a valid touch. For some touchpad with a low reporting rate, especially the PS2 touchpad perhaps, it might be hard to reach that when the click contact is too short.

The commit on VoodooInput mentioned above should solve the Smart Zoom issue. Maybe the commit in this PR is also required.

Reply:
> Hm, I tested before this PR and fingers are still sometimes left with 0 size, so it seems unrelated. Though without the fix it seems to be completely impossible to trigger Smart Zoom, and with the fix it works at least sometimes. It also fixes a problem in three-finger drag. I reimplemented only this part of this fix in [cfa2484](https://github.com/acidanthera/VoodooPS2/commit/cfa24845e87e2c886e48cff42b877103ec14504b). Thanks and sorry for misunderstanding!

I'm sorry I didn't express the key content in this pull request, which aim to fix the gesture issue of Notification Centre and Smart Zoom together.

The code in Line 1581, it is used to fix extra finger data, now it has been solved in VoodooInput. And other code are used to send release event when fingers leave, now it has been solved in VoodooPS2.

I have updated my local source code and tested the new VoodooInput and VoodooPS2 that you committed, and the Smart Zoom issue still exist.In order to solve it, @Goshin has showed the solution and it can work well now.I think you can check his commit and consider whether merging it into master branch.

Reply:
> For MT2Simulator, the `touch_state` changes are necessary for some gestures, but it needs to reach `touch_state = 4` to be recognized as a valid touch. For some touchpad with a low reporting rate, especially the PS2 touchpad perhaps, it might be hard to reach that when the click contact is too short.
> 
> The commit on VoodooInput mentioned above should solve the Smart Zoom issue. Maybe the commit in this PR is also required.
Yes, with your VoodooInput commit Smart Zoom works good, and even two-finger single tap works better than before. Are that changes appropriate for VoodooI2C, or it should be somehow done only for PS/2 touchpads?

Reply:
Five-finger gestures now also work almost always.

Reply:
Yep good for VoodooI2C too.